File: frontend/management/src/app/modules/admin/modules/products/model/fabric-composition-item.ts
import { FabricType } from './fabric-type';
import { validateProps } from '@kicherkrabbe/shared';

export class FabricCompositionItem {
  readonly fabricType: FabricType;
  readonly percentage: number;

  private constructor(props: { fabricType: FabricType; percentage: number }) {
    validateProps(props);

    this.fabricType = props.fabricType;
    this.percentage = props.percentage;
  }

  static of(props: { fabricType: FabricType; percentage: number }): FabricCompositionItem {
    return new FabricCompositionItem({
      fabricType: props.fabricType,
      percentage: props.percentage,
    });
  }
}

File: backend/features/offers/offers-core/src/main/java/de/bennyboer/kicherkrabbe/offers/FabricCompositionItem.java
package de.bennyboer.kicherkrabbe.offers;

import lombok.AllArgsConstructor;
import lombok.Value;

import static de.bennyboer.kicherkrabbe.commons.Preconditions.notNull;
import static lombok.AccessLevel.PRIVATE;

@Value
@AllArgsConstructor(access = PRIVATE)
public class FabricCompositionItem {

    FabricType fabricType;

    LowPrecisionFloat percentage;

    public static FabricCompositionItem of(FabricType fabricType, LowPrecisionFloat percentage) {
        notNull(fabricType, "Fabric type must be given");
        notNull(percentage, "Percentage must be given");

        return new FabricCompositionItem(fabricType, percentage);
    }

    @Override
    public String toString() {
        return "FabricCompositionItem(fabricType=%s, percentage=%s)".formatted(fabricType, percentage);
    }

}
